### PR TITLE
Allow tracing during startup

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

File: sky/shell/tracing_controller.cc [hunk 1]
```diff
@@ -97,6 +99,17 @@ TracingController::~TracingController() {
   ManageObservatoryCallbacks(false);
 }
 
+void TracingController::SetDartInitialized() {
+  dart_initialized_ = true;
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: ASSERT(!dart_initialized)?

File: sky/shell/tracing_controller.cc [hunk 2]
```diff
@@ -108,6 +121,8 @@ void TracingController::ManageObservatoryCallbacks(bool add) {
 void TracingController::StartTracing() {
   LOG(INFO) << "Starting trace";
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: ASSERT(!tracing_active)?

File: sky/shell/tracing_controller.cc [hunk 3]
```diff
@@ -138,7 +153,9 @@ void TracingController::StopTracingAsync(const base::FilePath& path,
 }
 
 void TracingController::StartDartTracing() {
-  Dart_GlobalTimelineSetRecordedStreams(~0);
+  if (dart_initialized_) {
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: I'd probably write this as early-return instead.  Allows the method to grow more later with ease.

if (!dart_initialized_)
  return;

File: sky/shell/tracing_controller.cc [hunk 4]
```diff
@@ -161,6 +178,12 @@ static void TracingController_DartStreamConsumer(
 }
 
 void TracingController::StopDartTracing() {
+  if (!dart_initialized_) {
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: Can this just early return?  It feels strange that StopDartTracing does more than just stop the dart bits. :(  Presumably we should split this method into "FinalizeTracing" and "StopDartTracing"?

File: sky/shell/ui/engine.cc [hunk 5]
```diff
@@ -78,6 +78,7 @@ void Engine::Init() {
   DCHECK(!g_platform_impl);
   g_platform_impl = new PlatformImpl();
   blink::initialize(g_platform_impl);
+  Shell::Shared().tracing_controller().SetDartInitialized();
```
Comment (eseidelGoogle) [inline sky/shell/ui/engine.cc]: Is this the right place to call this?  I guess blink::initialize calls InitVM?

File: sky/shell/tracing_controller.cc (hunk 2 above)
Comment (chinmaygarde) [inline sky/shell/tracing_controller.cc]: The user can press the start recording button multiple times in observatory. I would rather not assert

File: sky/shell/tracing_controller.cc (hunk 4 above)
Comment (chinmaygarde) [inline sky/shell/tracing_controller.cc]: Yeah, the early return is required.

File: sky/shell/tracing_controller.cc (hunk 2 above)
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: It feels like that concern should go next to the observatory handling bits.  Seems like it's a programming error (hence the assert) to call this StartTracing manually more than once?  Maybe not?  Donno.

Comment (chinmaygarde) [inline sky/shell/tracing_controller.cc]: ok

File: sky/shell/tracing_controller.cc [hunk 6]
```diff
@@ -25,6 +27,8 @@ const char kObservatoryMethodStopTracing[] = "flutter_stopTracing";
 const char kObservatoryResultOk[] = "{\"success\" : true}";
 const char kObservatoryResultFail[] = "{\"success\" : false}";
 
+const char kTracingSwitchEnableOnStartup[] = "startup-trace";
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: I probably would have called this "trace-startup":
https://code.google.com/p/chromium/codesearch#chromium/src/components/tracing/tracing_switches.cc&q=trace-startup&sq=package:chromium&type=cs&l=33

File: sky/shell/tracing_controller.cc [hunk 7]
```diff
@@ -41,46 +45,56 @@ static const char* ObservatoryInvoke(const char* method,
 
   if (strncmp(method, kObservatoryMethodStartTracing,
               sizeof(kObservatoryMethodStartTracing)) == 0) {
-    tracing_controller->StartTracing();
-    return strdup(kObservatoryResultOk);
+    if (!tracing_controller->tracing_active()) {
+      tracing_controller->StartTracing();
+      return strdup(kObservatoryResultOk);
+    } else {
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: No else needed after a return, unless that's our style?

File: sky/shell/tracing_controller.cc [hunk 8]
```diff
@@ -41,46 +45,56 @@ static const char* ObservatoryInvoke(const char* method,
...
-    DCHECK(length == length_read);
+      DCHECK(length == length_read);
 
-    base::DeleteFile(path, false);
+      base::DeleteFile(path, false);
 
-    return data;
+      return data;
+    } else {
+      return strdup(kObservatoryResultFail);
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: No else needed after return, as above.

File: sky/shell/tracing_controller.cc [hunk 9]
```diff
@@ -89,14 +103,33 @@ static const char* ObservatoryInvoke(const char* method,
 TracingController::TracingController()
     : picture_tracing_enabled_(false),
       terminate_loop_on_write_(false),
+      dart_initialized_(false),
+      tracing_active_(false),
       weak_factory_(this) {
   ManageObservatoryCallbacks(true);
+
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: Why not just do this in the caller?  Then this file doesn't need to know about commandline, etc.

Also, I think we have a switches.cc where these switches should be defined.

File: sky/shell/tracing_controller.cc [hunk 10]
```diff
@@ -2,14 +2,16 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+#include "sky/shell/tracing_controller.h"
+
+#include "base/command_line.h"
```
Comment (eseidelGoogle) [inline sky/shell/tracing_controller.cc]: No longer needed?

**Comment by eseidelGoogle** [comment]
There are still some unaddressed nits, which is OK, but please do a once-over read-through and consider.  Please squash before landing, don't need to see again.  LGTM.

**Comment by chinmaygarde** [comment]
Addressed. Rebased.

**Pull request merged**